core/state/snapshot/generate.go

```diff
@@ -404,7 +410,11 @@ func (dl *diskLayer) generateRange(root common.Hash, prefix []byte, kind string,
 		// The verification is passed, process each state with the given
 		// callback function. If this state represents a contract, the
 		// corresponding storage check will be performed in the callback
-		if err := result.forEach(func(key []byte, val []byte) error { return onState(key, val, false, false) }); err != nil {
+		var drange *DanglingRange
+		if kind != "storage" {
+			drange = NewDanglingRange(dl.diskdb, origin, result.last(), false)
```

Comment:
I wonder if `result.Last` can be wrong. Example: 

The first range is `0x000 - 0x00F`. It has the following : 
````
0x001: an account + storage
0x002: an account + storage
0x003: an account + storage
0x00a: an account + storage
```
Then most of them go away, and we're left with only one account in this range: 
```
0x001: account + storage
```
Won't `result.last` be `0x001`, making it so we don't scan `0x001 - 0x00F` ?

Reply:
Why do you say `Then most of them go away`? If we iterate out four accounts `0x001`, `0x002`, `0x003`, `0x00a`, then `result.last` will be `0x00a`. It represents the account snapshot in range `0x001 - 0x00a` is checked by generator. And the next range should start from `0x00b`.

Reply:
Even for these four accounts, three of them are not in trie(dangling account snapshot data) which will be removed from the disk, the `result.last` will still be `0x00a`.